### PR TITLE
Electrodialysis  Flowsheet 

### PR DESCRIPTION
**Pull request by lbibl**
issue #593 

## Summary/Motivation:
This is a PR in preparation for creating a flowsheet for electrodialysis desalination for a NaCl solution.  

## Changes proposed in this PR:
- Minor electrodialysis unit model refinement and modification may be included in this PR

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:42]: If you call this main function in the test file (i.e. test_main_function or something), all of this code will be covered and the display functions would be covered

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py]: typo: "inlects"

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py]: Why are you "building" these? Is it for initialization or displaying?

I understand why you them to be created before scaling. I usually have a comment that says "touching variables for initialization and displaying results" rather than "build extra properties as needed"

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py]: This could be replaced with our utility function: watertap.core.util.initilization.check_dof

We should probably update this for all flowsheets to centralize this check and error message

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:225]: The system already had set_operating_conditions called, at least in the workflow in main(). I'd recommend deleting this here.

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:243]: This comment says less than 1 g/L, but the constraint is less than 0.393 g/L right? Ohhh its Na vs NaCl. You could just add the concentration of Na to the concentration of Cl and set it.

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:242]: This 0.393 is a magic number, I'd recommend creating a parameter with this value and then using the parameter in the constraint.

```
m.fs.product.conc_mass_Na_max = Param(initialize=0.393, ...)
m.fs.eq_product_quality = Constraint(
        expr=m.fs.product.properties[0].conc_mass_phase_comp["Liq", "Na_+"] <= m.fs.product.conc_mass_Na_max
)
```

File: watertap/examples/flowsheets/electrodialysis/tests/test_electrodialysis_1stack.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/electrodialysis/tests/test_electrodialysis_1stack.py]: You don't need to test all of these things in the flowsheet test, if these were to fail the unit model tests would catch that.

You should try to only test the things that the flowsheet adds/does

**Review by TimBartholomew — APPROVED** [review]
I have several comments, I'm not sure it should hold up the PR, especially since I know you started work on other things on the same branch locally. But they should get addressed sometime.

**Review by adam-a-a — APPROVED** [review]
LGTM - since I know you intend to continue refining the ED model/flowsheet, I won't hold this back. However, I agree with Tim's suggested changes, and I think you should keep track of these things for a subsequent PR. 

Good job!

**Comment by adam-a-a** [comment]
@lbibl I'll enable auto-merge so long as you take one last look at the comments here to keep in mind for future PRs. Let me know when you're ready to merge.

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (lbibl) [reply watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:42]: Added a simple test to run main().

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (lbibl) [reply watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:225]: I kept this for now for cases where the main() is not called.  Will check this back in the future.

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (lbibl) [reply watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:243]: Did this.

File: watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (lbibl) [reply watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack.py:242]: I used the 1g/L for both ions in this version.  Will check for a better way as suggested in the future.

**Comment by lbibl** [comment]
Thank you to @TimBartholomew and @adam-a-a .   I addressed most of Tim's comments.  This should be good to go after passing checks.

File: watertap/unit_models/electrodialysis_0D.py
Comment (adam-a-a) [inline watertap/unit_models/electrodialysis_0D.py]: Why change this to 10?

File: watertap/unit_models/electrodialysis_0D.py
Comment (adam-a-a) [inline watertap/unit_models/electrodialysis_0D.py]: `recovery_mass_H2O` might be more consistent with the way we named water recovery elsewhere in WaterTAP. We also have `recovery_mass_phase_comp`, indexed by phase and component, as well as `recovery_vol_phase`.

**Review by adam-a-a — APPROVED** [review]
Approving again, but have 2 comments to consider.

File: watertap/unit_models/electrodialysis_0D.py
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0D.py]: I changed the name for mass.   I will add the other two when needed in the future.

File: watertap/unit_models/electrodialysis_0D.py
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0D.py]: Fixed. (thank you for catching this)

**Pull request merged**
